File: src/test/java/itmo/high_perf_sys/chat/service/SubscriptionServiceTest.java
package itmo.high_perf_sys.chat.service;

import itmo.high_perf_sys.chat.dto.customer.request.CreateUserAccountRequest;
import itmo.high_perf_sys.chat.dto.customer.request.UpdateUserInfoRequest;
import itmo.high_perf_sys.chat.dto.customer.response.GetUserInfoResponse;
import itmo.high_perf_sys.chat.dto.subs.request.CreateSubRequest;
import itmo.high_perf_sys.chat.dto.subs.response.SubscriptionResponse;
import itmo.high_perf_sys.chat.entity.*;
import itmo.high_perf_sys.chat.exception.UserAccountNotFoundException;
import itmo.high_perf_sys.chat.repository.SubRepository;
import itmo.high_perf_sys.chat.repository.UserRepository;
import itmo.high_perf_sys.chat.repository.chat.UsersChatsRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class SubscriptionServiceTest {

    private UserService userService;
    private SubRepository subRepository;
    private SubscriptionService subscriptionService;

    @BeforeEach
    void setUp() {
        userService = mock(UserService.class);
        subRepository = mock(SubRepository.class);
        subscriptionService = new SubscriptionService(userService, subRepository);
    }

    @Test
    public void testCreateSub() {
        UUID userId = UUID.randomUUID();
        UUID subscribedUserId = UUID.randomUUID();
        CreateSubRequest request = mock(CreateSubRequest.class);

        when(request.userId()).thenReturn(userId);
        when(request.subscribedUserId()).thenReturn(subscribedUserId);

        when(userService.existsById(userId)).thenReturn(true);
        when(userService.existsById(subscribedUserId)).thenReturn(true);

        UUID subId = UUID.randomUUID();
        Subscribers newSubscriber = new Subscribers();
        newSubscriber.setId(subId);

        when(subRepository.save(any(Subscribers.class))).thenReturn(newSubscriber);

        UUID result = subscriptionService.createSub(request);

        assertNotNull(result);
        assertEquals(subId, result);
        verify(userService, times(1)).existsById(userId);
        verify(userService, times(1)).existsById(subscribedUserId);
        verify(subRepository, times(1)).save(any(Subscribers.class));
    }

    @Test
    public void testCreateSubThrowsWhenUserNotFound() {
        UUID userId = UUID.randomUUID();
        UUID subscribedUserId = UUID.randomUUID();
        CreateSubRequest request = mock(CreateSubRequest.class);

        when(request.userId()).thenReturn(userId);
        when(request.subscribedUserId()).thenReturn(subscribedUserId);

        when(userService.existsById(userId)).thenReturn(false);

        assertThrows(UserAccountNotFoundException.class, () -> subscriptionService.createSub(request));
        verify(userService, times(1)).existsById(userId);
        verify(userService, never()).existsById(subscribedUserId);
        verify(subRepository, never()).save(any(Subscribers.class));
    }

    @Test
    void testGetSub() {
        UUID subId = UUID.randomUUID();
        Subscribers subscriber = new Subscribers();
        subscriber.setId(subId);

        when(subRepository.findById(subId)).thenReturn(Optional.of(subscriber));

        Subscribers result = subscriptionService.getSub(subId);

        assertNotNull(result);
        assertEquals(subId, result.getId());
        verify(subRepository, times(1)).findById(subId);
    }

    @Test
    void testGetSubThrowsWhenNotFound() {
        UUID subId = UUID.randomUUID();

        when(subRepository.findById(subId)).thenReturn(Optional.empty());

        assertThrows(IllegalArgumentException.class, () -> subscriptionService.getSub(subId));
        verify(subRepository, times(1)).findById(subId);
    }

    @Test
    void testDeleteSub() {
        UUID subId = UUID.randomUUID();

        when(subRepository.existsById(subId)).thenReturn(true);

        subscriptionService.deleteSub(subId);

        verify(subRepository, times(1)).existsById(subId);
        verify(subRepository, times(1)).deleteById(subId);
    }

    @Test
    void testDeleteSubThrowsWhenNotFound() {
        UUID subId = UUID.randomUUID();

        when(subRepository.existsById(subId)).thenReturn(false);

        assertThrows(IllegalArgumentException.class, () -> subscriptionService.deleteSub(subId));
        verify(subRepository, times(1)).existsById(subId);
        verify(subRepository, never()).deleteById(subId);
    }

    @Test
    void testGetSubscriptionsByUserId() {
        UUID userId = UUID.randomUUID();

        SubscriptionResponse response1 = new SubscriptionResponse(UUID.randomUUID(), UUID.randomUUID(), LocalDateTime.now());
        SubscriptionResponse response2 = new SubscriptionResponse(UUID.randomUUID(), UUID.randomUUID(), LocalDateTime.now());
        List<SubscriptionResponse> mockResponseList = List.of(response1, response2);

        when(subRepository.getSubResponseByUserId(userId)).thenReturn(mockResponseList);

        List<SubscriptionResponse> result = subscriptionService.getSubscriptionsByUserId(userId);

        assertNotNull(result);
        assertEquals(2, result.size());
        assertEquals(mockResponseList, result);
        verify(subRepository, times(1)).getSubResponseByUserId(userId);
    }
}

File: src/test/java/itmo/high_perf_sys/chat/service/SubscriptionServiceTest.kt
package itmo.high_perf_sys.chat.service

import itmo.high_perf_sys.chat.dto.subs.request.CreateSubRequest
import itmo.high_perf_sys.chat.dto.subs.response.SubscriptionResponse
import itmo.high_perf_sys.chat.entity.Subscribers
import itmo.high_perf_sys.chat.exception.UserAccountNotFoundException
import itmo.high_perf_sys.chat.repository.SubRepository
import itmo.high_perf_sys.chat.repository.UserRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import java.time.LocalDateTime
import org.junit.Assert.*
import java.util.*


class SubscriptionServiceTest {

    @Mock
    private lateinit var userRepository: UserRepository

    @Mock
    private lateinit var subscribersRepository: SubRepository

    @InjectMocks
    private lateinit var subscriptionService: SubscriptionService

    @BeforeEach
    fun setUp() {
        MockitoAnnotations.openMocks(this)
    }

    @Test
    fun testCreateSub() {
        val createSubRequest = CreateSubRequest(
            userId = UUID.randomUUID(),
            subscribedUserId = UUID.randomUUID()
        )

        `when`(userRepository.existsById(createSubRequest.userId)).thenReturn(true)
        `when`(userRepository.existsById(createSubRequest.subscribedUserId)).thenReturn(true)
        `when`(subscribersRepository.save(any())).thenAnswer { invocation ->
            val subscriber = invocation.getArgument<Subscribers>(0)
            subscriber.id = UUID.randomUUID()
            subscriber
        }

        val subId = subscriptionService.createSub(createSubRequest)

        assertNotNull(subId)
        verify(subscribersRepository, times(1)).save(any())
    }

    @Test
    fun testCreateSubUserNotFound() {
        val createSubRequest = CreateSubRequest(
            userId = UUID.randomUUID(),
            subscribedUserId = UUID.randomUUID()
        )

        `when`(userRepository.existsById(createSubRequest.userId)).thenReturn(false)

        assertFailsWith<UserAccountNotFoundException> { subscriptionService.createSub(createSubRequest) }
    }

    @Test
    fun testCreateSubSubscribedUserNotFound() {
        val createSubRequest = CreateSubRequest(
            userId = UUID.randomUUID(),
            subscribedUserId = UUID.randomUUID()
        )

        `when`(userRepository.existsById(createSubRequest.userId)).thenReturn(true)
        `when`(userRepository.existsById(createSubRequest.subscribedUserId)).thenReturn(false)

        assertFailsWith<UserAccountNotFoundException> { subscriptionService.createSub(createSubRequest) }
    }

    @Test
    fun testGetSub() {
        val subId = UUID.randomUUID()
        val subscriber = Subscribers().apply {
            id = subId
            userId = UUID.randomUUID()
            subscribedUserId = UUID.randomUUID()
            subscriptionTime = LocalDateTime.now()
        }

        `when`(subscribersRepository.findById(subId)).thenReturn(Optional.of(subscriber))

        val result = subscriptionService.getSub(subId)

        assertNotNull(result)
        assertEquals(subId, result.id)
    }

    @Test
    fun testGetSubNotFound() {
        val subId = UUID.randomUUID()

        `when`(subscribersRepository.findById(subId)).thenReturn(Optional.empty())

        assertFailsWith<IllegalArgumentException> { subscriptionService.getSub(subId) }
    }

    @Test
    fun testDeleteSub() {
        val subId = UUID.randomUUID()

        `when`(subscribersRepository.existsById(subId)).thenReturn(true)

        subscriptionService.deleteSub(subId)

        verify(subscribersRepository, times(1)).deleteById(subId)
    }

    @Test
    fun testDeleteSubNotFound() {
        val subId = UUID.randomUUID()

        `when`(subscribersRepository.existsById(subId)).thenReturn(false)

        assertFailsWith<IllegalArgumentException> { subscriptionService.deleteSub(subId) }
    }

    @Test
    fun testGetSubscriptionsByUserId() {
        val userId = UUID.randomUUID()
        val subscriptions = listOf(
            SubscriptionResponse(
                id = UUID.randomUUID(),
                userId = userId,
                subscribedUserId = UUID.randomUUID(),
                subscriptionTime = LocalDateTime.now()
            ),
            SubscriptionResponse(
                id = UUID.randomUUID(),
                userId = userId,
                subscribedUserId = UUID.randomUUID(),
                subscriptionTime = LocalDateTime.now()
            )
        )

        `when`(subscribersRepository.getSubResponseByUserId(userId)).thenReturn(subscriptions)

        val result = subscriptionService.getSubscriptionsByUserId(userId)

        assertNotNull(result)
        assertEquals(2, result.size)
    }
}
